bluetoe/link_layer/include/bluetoe/peripheral_latency.hpp

```diff
@@ -0,0 +1,318 @@
+ *
+ * @sa bluetoe::link_layer::peripheral_latency_ignored
+ * @sa ...
+ */
+namespace bluetoe {
+namespace link_layer {
+
+    namespace details {
+        struct peripheral_latency_meta_type {};
+    }
+
+    /**
+     * @brief detailed options for the peripheral latency behavior
+     *
+     * Every option defines a set of events / circumstances under which
+     * the link layer should listen to the very next connection event.
+     * (beside the minumum events, the link layer has to listen to due
+     * to the periperhal latency value of the connection).
+     *
+     * @note in combination, there are some options that make no sense.
+     *       For example, using listen_always with some of the other
+     *       listen_* options is pointless. Also combining send_always
+     *       with send_pending_data doesn not make sense.
+     */
+    enum class peripheral_latency
+    {
+        /**
+         * @brief listen at all connection events.
+         *
+         * This does not imply, that the link layer will also respond at every
+         * connection event.
+         */
+        listen_always,
+
+        /**
+         * @brief listen, if the link layer contains unacknowlaged data.
+         *
+         * If a PDU was send out, but no acknowlagement was jet received,
+         * the link layer will listen at the next connection event, if
+         * this options is set.
+         *
+         * This might be interesting, if the sending queue is very small because
+         * Bluetoe has to keep a copy of a transmitted PDU, as long as that PDU
+         * is not acknowlaged.
+         */
+        listen_if_unacknowlaged_data,
+
+        /**
+         * @brief listen to the next connection event, if the last received PDU was
+         *        not empty.
+         */
+        listen_if_last_received_not_empty,
+
+        /**
+         * @brief listen to the next connection event, if the last transmitted PDU was
+         *        not empty.
+         */
+        listen_if_last_transmitted_not_empty,
+
+        /**
+         * @brief listen to the next connection event, if the last received PDU had
+         *        the more data (MD) flag set.
+         *
+         * With the MD flag, the central (and the peripheral) can indicate, that there
+         * is more data to be send. Usually this data is then transmitted during a
+         * connection even. If there is no support for the MD flag either by the peripheral
+         * or by the central, then, this option might be usefull.
+         */
+        listen_if_last_received_had_more_data,
+
+        /**
+         * @brief listen to the next connection event, if for the last connection event,
+         *        a crc error happend, or a timeout occured.
+         *
+         * This behavior is required by the Bluetooth specification and should not left out
+         * without good reason.
+         */
+        listen_if_last_received_had_error,
```

Comment:
shouldn't this then be always there, rather than be optional ?

Reply:
Yes, if you design your application, you should include this option. If you have a very good reason to not do so, you can leave that out. The provided Bluetoe default will probably add this option.

Reply:
I think this should be optional because this is only a recommendation (should) and not a requirement (shall) in the spec (v5.3 Vol 6 Part B Section 4.5.1:
_"If the Peripheral does not receive a valid packet from the
Central after applying Peripheral latency, it **should** listen at each of the
subrated anchor points and not apply Peripheral latency until it receives a
packet from the Central."_

Another usage I see for making this configurable is if one wants to use Bluetoe as a test device, to simulate wrong behaviors (for robustness tests for example).

How about moving this one first in the list? Maybe it would make sense to make it crescendo: at first the default ones, then add the more and more extreme options (finishing with listen_always -- this will also help the reader better comprehend what listen_always does)